src/lib/support/BytesToHex.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "BytesToHex.h"
+
+namespace chip {
+namespace Encoding {
+
+namespace {
+
+char NibbleToHex(uint8_t nibble, bool uppercase)
+{
+    char x = static_cast<char>(nibble & 0xFu);
+
+    if (x >= 10)
+    {
+        return static_cast<char>((x - 10) + (uppercase ? 'A' : 'a'));
+    }
+    else
+    {
+        return static_cast<char>(x + '0');
+    }
+}
+
+} // namespace
+
+CHIP_ERROR BytesToHex(const uint8_t * src_bytes, size_t src_size, char * dest_hex, size_t dest_size_max, bool uppercase,
+                      bool nul_terminate)
+{
+    if ((src_bytes == nullptr) || (dest_hex == nullptr))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    size_t expected_output_size = (src_size * 2u) + (nul_terminate ? 1u : 0u);
```

Comment:
Might be worth a comment about why we're note worried about overflow here.

Or maybe we _should_ worry about it, and should rewrite the test accordingly:
```
if (nul_terminate) {
  if (dest_size_max / 2 < expected_output_size || 2*expected_output_size >= dest_size_max) {
    return CHIP_ERROR_BUFFER_TOO_SMALL;
  }
} else {
  if (dest_size_max / 2 < expected_output_size || 2*expected_output_size > dest_size_max) {
    return CHIP_ERROR_BUFFER_TOO_SMALL;
  }
}
```
or some such; I am not a big fan of the comprehensibility of that code....

Reply:
Found a clean way